src/Interpreters/Cache/FileCache.cpp

```diff
@@ -776,7 +782,17 @@ bool FileCache::tryReserve(
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheReserveMicroseconds);
 
     assertInitialized();
-    auto cache_lock = lockCache();
+    auto cache_lock = tryLockCache();
+    if (!cache_lock)
+    {
+        ProfileEvents::increment(ProfileEvents::FilesystemCacheReserveLockFailed);
```

Comment:
I think in case when we failed to acquire a lock we need to remember how much we wanted to reserve there (and accumulate this knowledge from all such failed cases). So the the thread which will get the lock will be able to reserve more space (let's say evict entries) for his friends.